Revamp SDL_IsTraySupported, add tray support check for GNOME

eafton opened the pull request:
-Revamps IsTraySupported, it now returns an "enum", allowing us to indicate if tooltip support is present and if tray support cannot be determined.
-Splits off GLib into its own module and adds GSettings support.
-Moves some common code into its own file (CheckCurrentDesktop).
-Adds tray extension check for GNOME desktops.
-Moves GLib and GTK hints into the hints header, adds hint for disabling aformentioned GNOME tray extension check.

Semphriss commented:
I've given a quick look to the API changes, and I'm wondering, what is the practical use of knowing the support status of system trays beyond yes/no?

I'm also unsure if the change breaks the ABI, I'll let someone more familiar than me verify that.

eafton commented:
> I've given a quick look to the API changes, and I'm wondering, what is the practical use of knowing the support status of system trays beyond yes/no?
> 
> I'm also unsure if the change breaks the ABI, I'll let someone more familiar than me verify that.

We can return unknown if the user is running an unrecognized desktop or if some kind of GLib error occurs, also the current GTK tooltip implementation does not support tootips unlike the OS X and Windows implementations, it would be good to let people know in advance.

API/ABI should not be broken since 3.4.0 hasnt been released yet.

Semphriss commented:
Oh, I didn't realize that main is now contains 3.4.0 stuff.

What's the use of knowing details about how trays are supported? Normally, the program either uses system trays, or it doesn't. The original function is already pretty much unneeded, since the expected workflow is to create a tray when the program starts, see if that fails, and deduce whether or not trays are supported based on that. I'm not sure what an app can do based on a return value other than yes and no (what should the program do if it's "unknown"? Try anyways? Don't try? In either of those case, would it be better to have something like "no/maybe,-you-should-try", so alias "unknown" to "yes"?).

eafton commented:
> Oh, I didn't realize that main is now contains 3.4.0 stuff.
> 
> What's the use of knowing details about how trays are supported? Normally, the program either uses system trays, or it doesn't. The original function is already pretty much unneeded, since the expected workflow is to create a tray when the program starts, see if that fails, and deduce whether or not trays are supported based on that. I'm not sure what an app can do based on a return value other than yes and no (what should the program do if it's "unknown"? Try anyways? Don't try? In either of those case, would it be better to have something like "no/maybe,-you-should-try", so alias "unknown" to "yes"?).

I would have moved the checks into CreateTray if this function did not exist, but it does, thats why I chose to expand it. I guess having IsTraySupported reduces overhead and allows the application to prepare an alternate user interface in advance if no tray support exists.

As for what an application should do if support is unknown, they should probably try but with a warning message.

Also, the tooltips flag was added because the current GTK tray implementation does not support it while the implementations for OS X and Windows do, I think having a flag for that is more readable instead of ifdef checking the platform. Also, I am working on a "generic" tray implementation does support tooltips that will live alongside the GTK implementation, thats another reason for the flag.

slouken commented:
Honestly, I'm not sure why SDL_IsTraySupported() was added. The flow should be:
* Try to create a tray
* If that fails, use an alternative UI for tray functionality if needed

This flow still needs to be done even if SDL_IsTraySupported() exists, so... ?

If there isn't a compelling reason to keep it, I'm planning to remove it.

eafton commented:
> @eafton, I'm not opposed to the other changes, but can you walk through them and explain why they're necessary?

Basically, GNOME, Panthenon and a few other DEs no longer support trays out of the box, I added those checks to make sure the user has a tray shell extension enabled on those desktops before attempting to create a tray item (there is a hint for bypassing this check if the user has an tray extension not recognized by the code), for this I needed to use GSettings, so I split off the GLib code and added GSettings support.

I also moved the GTK/GLib hints into the hints header so they could be documented.

One last thing, do I add an TrayAreTooltipsSupported function as a partial replacement?

slouken commented:
> > @eafton, I'm not opposed to the other changes, but can you walk through them and explain why they're necessary?
> 
> Basically, GNOME, Panthenon and a few other DEs no longer support trays out of the box, I added those checks to make sure the user has a tray shell extension enabled on those desktops before attempting to create a tray item (there is a hint for bypassing this check if the user has an tray extension not recognized by the code), for this I needed to use GSettings, so I split off the GLib code and added GSettings support.

I'm hesitant to load Gtk/Glib into SDL in the general case. We're looking at removing this for the DPI settings as well. Is this something you can query via dbus instead of pulling in all of GSettings?

> I also moved the GTK/GLib hints into the hints header so they could be documented.

See above ;)

> One last thing, do I add an TrayAreTooltipsSupported function as a partial replacement?

No, I think it's harmless to have someone set a tooltip even if the system has no way to display it. We do that for optional parts of dialogs as well.

eafton commented:
> I'm hesitant to load Gtk/Glib into SDL in the general case. We're looking at removing this for the DPI settings as well. Is this something you can query via dbus instead of pulling in all of GSettings? 
Actually I was thinking of working on a tray implementation based on that toolkit PR and just scrapping the GTK implementation completely (I am working on Wayland counterpart to that too, based on SDL_Surfaces and a built in basic bitmap font with optional libfcft support, but thats a topic for another day) but yes, you sadly need to use GSettings if you need to query the list of active shell extensions.

> No, I think it's harmless to have someone set a tooltip even if the system has no way to display it. We do that for optional parts of dialogs as well.

Ah, understood, I was actually thinking of adding a function to query what dialog options are supported (because I saw the portal implementation does not support things like changing the text of the cancel button, etc), but I guess I wont do that.
